state-sync/src/coordinator.rs

```diff
@@ -1113,6 +1076,118 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
         })
     }
 
+    fn verify_chunk_response_is_valid(
+        &mut self,
+        peer: &PeerNetworkId,
+        response: &GetChunkResponse,
+    ) -> Result<(), Error> {
+        // Verify response comes from upstream peer
+        if !self.request_manager.is_known_upstream_peer(peer) {
+            counters::RESPONSE_FROM_DOWNSTREAM_COUNT
+                .with_label_values(&[
+                    &peer.raw_network_id().to_string(),
+                    &peer.peer_id().to_string(),
+                ])
+                .inc();
+            self.request_manager.process_chunk_from_downstream(&peer);
+            return Err(Error::ReceivedChunkFromDownstream(peer.to_string()));
+        }
+
+        // Verify the chunk is not empty and that it starts at the correct version
+        if let Some(first_chunk_version) = response.txn_list_with_proof.first_transaction_version {
+            let known_version = self.local_state.synced_version();
+            let expected_version = known_version
+                .checked_add(1)
+                .ok_or_else(|| Error::IntegerOverflow("Expected version has overflown!".into()))?;
+
+            if first_chunk_version != expected_version {
+                self.request_manager.process_chunk_version_mismatch(
+                    &peer,
+                    first_chunk_version,
+                    known_version,
+                )?;
+            }
+        } else {
+            // The chunk is empty
+            self.request_manager.process_empty_chunk(&peer);
+            return Err(Error::ReceivedEmptyChunk(peer.to_string()));
+        }
+
+        // Verify the chunk has the expected type for the current syncing mode
+        match response.response_li.clone() {
+            ResponseLedgerInfo::LedgerInfoForWaypoint {
+                waypoint_li,
+                end_of_epoch_li,
+            } => {
+                if self.is_initialized() {
```

Comment:
any reason that self.sync_request.is_some isn't here? feels weird seeing that slight asymmetry

Reply:
State sync shouldn't ever get into the state where self.is_initialized() is false, and sync_request.is_some() is true. However, for completeness and "hardness", I've added the check.